### PR TITLE
Support for persistence to EventStoreDB. Fixes #3

### LINKED ISSUE #3: Add Event Store persistence layer

### PR DESCRIPTION
**Pull request by jonclare**

### CODE REVIEW

File: src/DomainLib.Persistence.EventStore/EventStoreEventDataExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿using EventStore.ClientAPI;
+
+namespace DomainLib.Persistence.EventStore
+{
+    public static class EventStoreEventDataExtensions
+    {
+        public static EventData ToEventData(this IEventSerializer @eventSerializer, object @event, string eventName)
+        {
+            var eventPersistenceData = eventSerializer.GetPersistenceData(@event, eventName);
+            return new EventData(eventPersistenceData.EventId, eventPersistenceData.EventName, eventPersistenceData.IsJsonBytes, eventPersistenceData.EventData, eventPersistenceData.EventMetadata);
```
Comment (carlosrfernandez) [inline src/DomainLib.Persistence.EventStore/EventStoreEventDataExtensions.cs:10]: Dude wrap this 😛

File: src/DomainLib/Aggregates/ApplyEventRouter.cs [hunk 2]
```diff
@@ -14,11 +14,15 @@ public class ApplyEventRouter<TAggregateRoot, TDomainEventBase>
     {
         private readonly IReadOnlyDictionary<Type, ApplyEvent<TAggregateRoot, TDomainEventBase>> _routes;
 
-        public ApplyEventRouter(IEnumerable<KeyValuePair<Type, ApplyEvent<TAggregateRoot, TDomainEventBase>>> routes)
+        public ApplyEventRouter(IEnumerable<KeyValuePair<Type, ApplyEvent<TAggregateRoot, TDomainEventBase>>> routes, 
+                                IEventNameMapping eventNameMappings)
         {
             _routes = ImmutableDictionary.CreateRange(routes);
+            EventNameMappings = eventNameMappings;
         }
```
Comment (daniel-smith) [inline src/DomainLib/Aggregates/ApplyEventRouter.cs:23]: Can we rename the interface to `IEventNameMap`, and the implementation to `EventNameMap`? Then the property here can also be called `EventNameMap`. Seems a bit more concise and consistent.

File: src/DomainLib.Persistence.EventStore/EventStoreEventsRepository.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        public EventStoreEventsRepository(IEventStoreConnection connection, IEventSerializer serializer)
+        {
+            _connection = connection;
+            _serializer = serializer;
+        }
+
+        public async Task<long> SaveEventsAsync<TEvent>(string streamName, long expectedStreamVersion, IEnumerable<TEvent> events)
+        {
+            var eventDatas = events.Select(e => _serializer.ToEventData(e));
+
```
Comment (daniel-smith) [inline src/DomainLib.Persistence.EventStore/EventStoreEventsRepository.cs:24]: Should we have an issue to address this TODO?

File: src/DomainLib/Aggregates/EventNameMapping.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        public void Merge(IEventNameMapping other, bool throwOnConflict = true)
+        {
+            foreach (var (key, value) in other.GetNameToTypeMappings())
+            {
+                RegisterEventName(key, value, throwOnConflict);
+            }
+        }
+
+        private static string DetermineEventNameForClrType(Type clrType)
+        {
```
Comment (daniel-smith) [inline src/DomainLib/Aggregates/EventNameMapping.cs]: Should this comment be part of a proper documentation comment somewhere?

**Review by daniel-smith — CHANGES_REQUESTED** [review]

File: src/DomainLib/Aggregates/ApplyEventRouter.cs (hunk 2 above)
Comment (jonclare) [reply src/DomainLib/Aggregates/ApplyEventRouter.cs:23]: Renamed

File: src/DomainLib.Persistence.EventStore/EventStoreEventsRepository.cs (hunk 3 above)
Comment (jonclare) [reply src/DomainLib.Persistence.EventStore/EventStoreEventsRepository.cs:24]: Created https://github.com/daniel-smith/domain-lib/issues/8

File: src/DomainLib/Aggregates/EventNameMapping.cs (hunk 4 above)
Comment (jonclare) [reply src/DomainLib/Aggregates/EventNameMapping.cs]: I'll set up a proper wiki site where we can document usage properly.

**Review by daniel-smith — APPROVED** [review]

**Pull request merged**